src/create_index.js

```diff
@@ -0,0 +1,89 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const request = require('request');
+const papa = require('papaparse');
+const lunr = require('lunr');
+
+const BASE_URL = 'http://www.indicepa.gov.it/public-services/opendata-read-service.php?dstype=FS&filename='
+const PUB_AMM_URL = `${BASE_URL}amministrazioni.txt`;
+const PEC_URL = `${BASE_URL}pec.txt`;
+
+const COD_AMM = 0;
+const DES_AMM = 1;
+const COMUNE = 2;
+const NOME_RESP = 3;
+const COGN_RESP = 4;
+const CAP = 5;
+const PROVINCIA = 6;
+const REGIONE = 7;
+const SITO_ISTITUZIONALE = 8;
+const INDIRIZZO = 9;
+const TITOLO_RESP = 10;
+const TIPOLOGIA_ISTAT = 11;
+const TIPOLOGIA_AMM = 12;
+const ACRONIMO = 13;
+const CF_VALIDATO = 14;
+const CF = 15;
+const MAIL1 = 16;
+const TIPO_MAIL1 = 17;
+const MAIL2 = 18;
+const TIPO_MAIL2 = 19;
+const MAIL3 = 20;
+const TIPO_MAIL3 = 22;
+const MAIL4 = 23;
+const TIPO_MAIL4 = 24;
+const MAIL5 = 25;
+const TIPO_MAIL5 = 26;
+const URL_FACEBOOK = 27;
+const URL_TWITTER = 28;
+const URL_GOOGLEPLUS = 29;
+const URL_YOUTUBE = 30;
+const LIV_ACCESSIBILI = 31;
+
+
+console.log(`GET: ${PUB_AMM_URL}`);
+request(PUB_AMM_URL, (err, res, csv_file) => {
+    if (err) { return console.log(err); }
+
+    console.log('converting the CSV string in an array');
+    const data = papa.parse(csv_file).data;
+    data.shift();
+
+    function Authority(row) {
+        this.code = row[COD_AMM],
+        this.description = row[DES_AMM]
+    }
+
+
+    console.log('index data in lunar.js format')
+    let paDb = {};
+    let index = lunr(function () {
+        this.ref('code');
+        this.field('code', { boost: 6 });
+        this.field('description', { boost: 3 });
+
+        data.forEach((row) => {
+            this.add({
+                code: row[COD_AMM],
+                description: row[DES_AMM]
+            });
+
+            paDb[row[COD_AMM]] = {
+                description: row[DES_AMM],
+                pec: row[TIPO_MAIL1] == 'pec' ? row[MAIL1] : ''
+            };
+
+        });
+    });
+
+
+    console.log('WRITE: authorities.index.json');
+    const serializedIndex = JSON.stringify(index);
+    fs.writeFileSync(path.join(__dirname, '..', 'public', 'assets', 'data', 'authorities.index.json'), serializedIndex);
```

Comment:
Forse meglio usare https://www.npmjs.com/package/write-file-atomic per evitare race condition? Oppure non succede perché node è single thread? Scusami ma conosco poco la piattaforma, ma dubito di default ogni volta che si aggiorna un file in modo non atomico.

Reply:
Node è single thread ma le operazioni di I/O avvengono su [thread separati](https://en.wikipedia.org/wiki/Libuv). Nella nostra applicazione, tuttavia, la scrittura di ogni file viene chiamata una sola volta all'avvio. L'unico scenario che mi viene in mente in cui ci potrebbe essere una race condition è quello in cui l'applicazione venga avviata due volte. Vedo che la libreria write-file-atomic scrive inizialmente su un altro file e poi cambia il nome se tutto va bene. Questo ci garantirebbe che se mentre si sta scrivendo il file, un client fa una richiesta di tale file, nella peggiore delle ipotesi il client riceverà un file più vecchio e mai un file vuoto. Considerando che i  file sono piccoli e quindi il tempo di scrittura è davvero breve e considerando che la scrittura del file avviene in parallelo all'avvio del server HTTP, quello che ho appena scritto è uno scenario molto remoto. Fatta questa lunga premessa :), vedo che integrare write-file-atomic è molto semplice quindi lo integrerei comunque per avere più stabilità.